Fix failing / flaky Navigation block e2e test by removing unnecessary validation steps for creation of draft page

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

The test `allows pages to be created from the navigation block and their links added to menu` from the Navigation block e2e tests is (still) intermittently failing.

We tried to fix this in https://github.com/WordPress/gutenberg/pull/37501 but it appears that was only half the battle!

Running locally it appears we've got some additional steps which attempt to validate that a draft page has been created that might not be necessary. 

I believe the behaviour of the Nav block no longer requires that we **re**select the Nav item in order to show the link preview. 

I've removed that step and it seems to pass. 

Fixes https://github.com/WordPress/gutenberg/issues/37479

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

We need to re-run CI several times (let's say x5) and check that the test is no longer flaky.

Please feel free to mark test runs below:

- [x] [Test round 1.](https://github.com/WordPress/gutenberg/actions/runs/1663782841/attempts/1)
- [x] [Test round 2](https://github.com/WordPress/gutenberg/actions/runs/1663782841/attempts/2).
- [x] [Test round 3](https://github.com/WordPress/gutenberg/actions/runs/1663782841/attempts/3).
- [x] [Test round 4](https://github.com/WordPress/gutenberg/actions/runs/1663782841/attempts/4).
- [x] Test round 5.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Remove unnecessary step before validating link is created

Mamaduka commented:
Failed once but because of the different Navigation test: `shows a warning if user does not have permission to edit or update navigation menus`.

talldan commented:
I don't think the fix in #37501 is correct. Using a mock in this way effectively changes behavior of the application and makes this test have less value. The test is no longer testing what happens in the real world.

The request does 404 reliably in production, so I think the test should expect that to happen rather than change the way the endpoint works to return an empty array.

I'm wondering why `expect( console ).toHaveErrored();` fails to work. That's unusual. Perhaps the test is completing before the error has been thrown. That's originally why I added the waitForNetworkIdle, but maybe that isn't working.

getdave commented:
> I don't think the fix in #37501 is correct. Using a mock in this way effectively changes behavior of the application and makes this test have less value. The test is no longer testing what happens in the real world.

I'm not sure about that one. The test is failing because of a 404 _in the consol_e. That isn't an _error_ with the Gutenberg application code. Rather it's a network issue. 

Yes this 404's in production but that's ok because that is handled gracefully in the UI. Check it in production and you'll see that happening.

All we're going is mocking an endpoint that's largely irrelevant to this test to avoid an unreliable 404 in the console causing a test failure.

Bottom line - network 404 != bug in application code.

getdave commented:
Also worth noting that with this fix (and the other one) in place we now have had this flaky test pass ✅  5 times.

talldan commented:
> Yes this 404's in production but that's ok because that is handled gracefully in the UI.

Yes, that's exactly my point. It should be handled gracefully in the UI and we should have tests in place to assert that this works. But we've now lost test coverage of this, because the endpoint had been modified to return an array in the end to end test.

getdave commented:
> It should be handled gracefully in the UI and we should have tests in place to assert that this works. But we've now lost test coverage of this, because the endpoint had been modified to return an array in the end to end test.

I understand but...this test isn't asserting on handling the rich previews. It's purely dealing with page creation in the Nav block. My thought was that we should have _another_ test that explicitly checks for graceful handling of rich previews. We're only mocking what is not directly under test.

> That's originally why I added the waitForNetworkIdle,

I think if we're resorting to adding workarounds to wait for network conditions then we're probably not constructing the test correctly. Network condition variability on CI is very often what causes flaky tests. 


...but maybe that isn't working.

Likely because the rich previews API will only 404 once the response comes back from the network. We were expecting a 404 in the test which only sometimes occurred because the rich preview endpoint is not what is under test.

talldan commented:
> I understand but...this test isn't asserting on handling the rich previews. It's purely dealing with page creation in the Nav block. My thought was that we should have another test that explicitly checks for graceful handling of rich previews. We're only mocking what is not directly under test.

This is true, though it is unfortunately part of the flow being tested, so it makes sense to write the test correctly.

The change still goes against the purpose of response mocking, which should be to provide realistic data. Returning an empty array here is modifying the runtime behavior of Gutenberg from testing code.

> I think if we're resorting to adding workarounds to wait for network conditions then we're probably not constructing the test correctly. Network condition variability on CI is very often what causes flaky tests.

I don't really see it as a workaround. We need to wait for the 404 response to happen before asserting that it happens, `waitForNetworkIdle` seemed the easiest way to do that, but it seems it doesn't have the intended effect.

getdave commented:
Thanks for the feedback and discussion. It's good to challenge my assumptions here.

As this is failing multiple PRs I'd like to move towards a solution.

My question now is: what would you like to see changed for you to feel comfortable with this PR?

Is it:

1. Removing the mock and finding a way to somehow make waiting for a 404 and an idle Network reliable and consistent?

_**OR**_

2. Leaving the mock but returning _realistic_ data (i.e. not an empty array) in order that we can feel comfortable the flow is adequately tested?

I'd vote for `#2` as it's more likely to produce a reliable test. I'm even happy to write new tests (in a follow up PR) to explicitly test the rich previews flow if that would help.

kevin940726 commented:
I agree with @talldan here that we shouldn't use unrealistic mock data and that `waitForNetworkIdle` is not a workaround, but I also agree with @getdave here that it's okay to mock data as long as it's not something related to the test itself.

The _ideal_ solution would either be:

1. Removing the mock and finding a way to somehow make waiting for a 404 and an idle Network reliable and consistent.
2. Keep the (realistic) mock, but also add tests for the 404 failures.

As for now, if this is blocking us, I agree to merge this as is (maybe with comments) to unblock us, but also create an issue to track this.

getdave commented:
One thing I just realised is that this test is skipped so it's less urgent than I thought 😄 

Still...it's good to not have disabled tests for too long.

getdave committed:
Use "real" mock data in response

Addresses https://github.com/WordPress/gutenberg/pull/37755#issuecomment-1009803740

getdave commented:
@talldan Now using appropriate mock data here.

talldan commented:
I still don't think that's right. The endpoint would normally return a payload like that, and the test doesn't ascertain that the front-end code handles the 404 from url-details. I think it'd be good to try making the test work properly without injecting incorrect data.

I had a go in https://github.com/WordPress/gutenberg/pull/37901, it passed consistently locally, so hopefully it does the same on CI 🤞 .

BTW - has there been a discussion about whether a 404 is the correct thing for url-details to return? My take is that it seems incorrect. The `url-details` resource could be found, so it should return a success code.

That the server couldn't find any details about the url is unrelated, that should be relayed via the response payload.

getdave commented:
Happy enough to close this in favour of https://github.com/WordPress/gutenberg/pull/37901.
